### PR TITLE
feat(experiment): accept async tarball_url_redirect function

### PR DESCRIPTION
**Pull request by tmota900**
PR to able to generate a tarball redirect url asynchronously. ref https://github.com/verdaccio/verdaccio/pull/1688

usage example:
```javascript
  experiments: {
    tarball_url_redirect: async (packageName, filename) => {
      return aws.createPresignedUrl({
        Key: `verdaccio/${packageName}/${filename}`,
      });
    },
  }
``` 
`config.js`

### CODE REVIEW

File: src/api/endpoint/api/package.ts [hunk 1]
```diff
@@ -45,12 +45,14 @@ const redirectOrDownloadStream = (
...
     .hasLocalTarball(packageName, filename)
-    .then((hasLocalTarball) => {
+    .then(async (hasLocalTarball) => {
       if (hasLocalTarball) {
         const context = { packageName, filename };
         const tarballUrl =
           typeof tarballUrlRedirect === 'function'
-            ? tarballUrlRedirect(context)
+            ? tarballUrlRedirect.constructor.name === 'AsyncFunction'
+              ? await tarballUrlRedirect(context)
```
Comment (juanpicado) [inline src/api/endpoint/api/package.ts:54]: If this fails, will crash the server  due there is no catch, I'm not sure about this approach mix thenable with async functions. I'm thinking how we can improve this.

Comment (juanpicado) [reply src/api/endpoint/api/package.ts]: Thanks, I'll check when I have some time 🙏🏼

**Comment by magalhas** [comment]
I would love to see this merged in order to enable an use case in the product I'm working on. Any status update on it? I see there's a label `6.x branch`. Any ETA?

**Comment by juanpicado** [comment]
That's the intentions, but working in other higher priorities.

**Review by juanpicado — APPROVED** [review]

**Comment by juanpicado** [comment]
Note: this goes to the branch 6.x if you want this changes right away for v5 versions we also can do that., please open a new PR and cherry pick the content.

**Pull request merged**